synapse/federation/transport/client.py

```diff
@@ -970,6 +970,27 @@ def write(self, data: bytes) -> int:
         return len(data)
 
     def finish(self) -> StateRequestResponse:
-        for c in self._coros:
-            c.close()
+        _close_coros(self._coros)
         return self._response
+
+
+def _close_coros(coros: Iterable[Generator[None, bytes, None]]) -> None:
+    """Close each of the given coroutines.
+
+    Always calls .close() on each coroutine, even if doing so raises an exception.
+    Any exceptions raised are aggregated into an ExceptionBundle.
+
+    :raises ExceptionBundle: if at least one coroutine fails to close.
+    """
+    exceptions = []
+    for c in coros:
+        try:
+            c.close()
+        except Exception as e:
+            exceptions.append(e)
+
+    if exceptions:
+        # raise from the first exception so that the traceback has slightly more context
+        raise ExceptionBundle(
+            f"There were {len(exceptions)} errors closing coroutines", exceptions
+        ) from exceptions[0]
```

Comment:
Not promising this is the cleanest way to do it---there are no tracebacks for `exceptions[1:]`---but this seemed Good Enough.

Reply:
(I'm not sure what @richvdh meant by exception chanining in https://github.com/matrix-org/synapse/issues/13293#issuecomment-1187643041---something along these lines?)